File: engine/strategies.py
"""
Six-strategy signal scanner. Each strategy returns a standardized dict:
  {signal, actionable, strength, ...strategy-specific fields}

Strength is a float 0-1 indicating how extreme the signal is.
"""

import numpy as np
from .indicators import sma, ema_array, compute_adx, compute_atr


def strategy_momentum_sma(closes):
    if len(closes) < 201:
        return {"strategy": "MOMENTUM_SMA", "signal": "INSUFFICIENT_DATA", "actionable": False, "strength": 0}
    fast = sma(closes, 50)
    slow = sma(closes, 200)
    fast_prev = sma(closes[:-1], 50)
    slow_prev = sma(closes[:-1], 200)

    spread = (fast - slow) / slow * 100

    if fast > slow and fast_prev <= slow_prev:
        sig, strength = "BUY", 0.9
    elif fast < slow and fast_prev >= slow_prev:
        sig, strength = "SELL", 0.9
    elif fast > slow:
        sig, strength = "BULLISH", min(spread / 10, 0.7)
    else:
        sig, strength = "BEARISH", min(abs(spread) / 10, 0.7)

    return {
        "strategy": "MOMENTUM_SMA", "signal": sig, "actionable": sig in ("BUY", "SELL"),
        "strength": round(strength, 2), "sma50": round(fast, 2),
        "sma200": round(slow, 2), "spread_pct": round(spread, 2),
    }


def strategy_ema18(closes):
    if len(closes) < 25:
        return {"strategy": "EMA18_BREAKOUT", "signal": "INSUFFICIENT_DATA", "actionable": False, "strength": 0}
    ema18 = ema_array(closes, 18)
    curr = closes[-1]
    prev = closes[-2]
    ema_today = ema18[-1]
    ema_yesterday = ema18[-2]
    distance = (curr - ema_today) / ema_today * 100

    if prev < ema_yesterday and curr > ema_today:
        sig, strength = "BUY", 0.8
    elif prev > ema_yesterday and curr < ema_today:
        sig, strength = "SELL", 0.8
    elif curr > ema_today:
        sig, strength = "ABOVE_EMA", min(distance / 5, 0.5)
    else:
        sig, strength = "BELOW_EMA", min(abs(distance) / 5, 0.5)

    return {
        "strategy": "EMA18_BREAKOUT", "signal": sig, "actionable": sig in ("BUY", "SELL"),
        "strength": round(strength, 2), "ema18": round(ema_today, 2),
        "distance_pct": round(distance, 2),
    }


def strategy_gold_impulse(closes):
    if len(closes) < 40:
        return {"strategy": "GOLD_IMPULSE", "signal": "INSUFFICIENT_DATA", "actionable": False, "strength": 0}
    ema13 = ema_array(closes, 13)
    ema12 = ema_array(closes, 12)
    ema26 = ema_array(closes, 26)

    min_len = min(len(ema12), len(ema26))
    macd_line = ema12[-min_len:] - ema26[-min_len:]
    signal_line = ema_array(macd_line, 9)

    if len(signal_line) < 2 or len(ema13) < 2:
        return {"strategy": "GOLD_IMPULSE", "signal": "INSUFFICIENT_DATA", "actionable": False, "strength": 0}

    hist_len = min(len(macd_line), len(signal_line))
    macd_hist = macd_line[-hist_len:] - signal_line[-hist_len:]

    ema_rising = ema13[-1] > ema13[-2]
    ema_falling = ema13[-1] < ema13[-2]
    hist_rising = macd_hist[-1] > macd_hist[-2]
    hist_falling = macd_hist[-1] < macd_hist[-2]

    if ema_rising and hist_rising:
        sig, strength = "BUY", 0.75
    elif ema_falling and hist_falling:
        sig, strength = "SELL", 0.75
    else:
        sig, strength = "NEUTRAL", 0.2

    return {
        "strategy": "GOLD_IMPULSE", "signal": sig, "actionable": sig in ("BUY", "SELL"),
        "strength": round(strength, 2), "macd_hist": round(float(macd_hist[-1]), 4),
    }


def strategy_8020(opens, highs, lows, closes):
    if len(closes) < 3:
        return {"strategy": "STRATEGY_8020", "signal": "INSUFFICIENT_DATA", "actionable": False, "strength": 0}

    body_prev = abs(closes[-2] - opens[-2])
    range_prev = highs[-2] - lows[-2]
    if range_prev == 0:
        return {"strategy": "STRATEGY_8020", "signal": "NO_SETUP", "actionable": False, "strength": 0}
    body_ratio = body_prev / range_prev

    if body_ratio < 0.80:
        return {"strategy": "STRATEGY_8020", "signal": "NO_SETUP", "actionable": False,
                "strength": 0, "body_ratio": round(body_ratio * 100, 1)}

    is_bullish = closes[-2] > opens[-2]
    is_bearish = closes[-2] < opens[-2]

    if is_bullish and closes[-1] < lows[-2]:
        sig, strength = "SELL", 0.85
    elif is_bearish and closes[-1] > highs[-2]:
        sig, strength = "BUY", 0.85
    elif is_bullish:
        sig, strength = "BEARISH_SETUP", 0.5
    elif is_bearish:
        sig, strength = "BULLISH_SETUP", 0.5
    else:
        sig, strength = "SETUP_PENDING", 0.3

    return {
        "strategy": "STRATEGY_8020", "signal": sig, "actionable": sig in ("BUY", "SELL"),
        "strength": round(strength, 2), "body_ratio": round(body_ratio * 100, 1),
    }


def strategy_volume(volumes, closes):
    if len(volumes) < 21:
        return {"strategy": "VOLUME", "signal": "INSUFFICIENT_DATA", "actionable": False, "strength": 0}
    avg_vol = np.mean(volumes[-21:-1])
    today_vol = volumes[-1]
    ratio = today_vol / avg_vol if avg_vol > 0 else 0
    price_change = (closes[-1] - closes[-2]) / closes[-2] * 100 if len(closes) >= 2 else 0

    if ratio > 2.5 and price_change > 1:
        sig, strength = "BREAKOUT_VOLUME", 0.9
    elif ratio > 2.5 and price_change < -1:
        sig, strength = "SELLOFF_VOLUME", 0.9
    elif ratio > 1.5 and price_change > 0.5:
        sig, strength = "HIGH_BULLISH", 0.6
    elif ratio > 1.5 and price_change < -0.5:
        sig, strength = "HIGH_BEARISH", 0.6
    elif ratio < 0.5:
        sig, strength = "DRY_UP", 0.3
    else:
        sig, strength = "NORMAL", 0.1

    return {
        "strategy": "VOLUME", "signal": sig, "actionable": sig in ("BREAKOUT_VOLUME", "SELLOFF_VOLUME"),
        "strength": round(strength, 2), "vol_ratio": round(ratio, 2),
        "price_change_pct": round(price_change, 2),
    }


def strategy_adx(highs, lows, closes):
    adx, plus_di, minus_di = compute_adx(highs, lows, closes, 14)
    if adx is None:
        return {"strategy": "ADX", "signal": "INSUFFICIENT_DATA", "actionable": False, "strength": 0}

    if adx > 25:
        regime = "TRENDING"
        sig = "TREND_BUY" if plus_di > minus_di else "TREND_SELL"
        strength = min(adx / 50, 1.0)
    elif adx < 20:
        regime = "RANGING"
        sig, strength = "NO_TREND", 0.1
    else:
        regime = "WEAK_TREND"
        sig = "LEAN_BULLISH" if plus_di > minus_di else "LEAN_BEARISH"
        strength = 0.3

    return {
        "strategy": "ADX", "signal": sig, "actionable": sig in ("TREND_BUY", "TREND_SELL"),
        "strength": round(strength, 2), "adx": adx, "plusDI": plus_di, "minusDI": minus_di,
        "regime": regime,
    }


def run_all_strategies(df):
    opens = df["Open"].values
    highs = df["High"].values
    lows = df["Low"].values
    closes = df["Close"].values
    volumes = df["Volume"].values

    results = {
        "momentum_sma": strategy_momentum_sma(closes),
        "ema18_breakout": strategy_ema18(closes),
        "gold_impulse": strategy_gold_impulse(closes),
        "strategy_8020": strategy_8020(opens, highs, lows, closes),
        "volume": strategy_volume(volumes, closes),
        "adx": strategy_adx(highs, lows, closes),
    }

    atr = compute_atr(highs, lows, closes, 14)
    results["_meta"] = {
        "last_price": closes[-1],
        "atr": round(atr, 2) if atr else None,
        "atr_pct": round(atr / closes[-1] * 100, 2) if atr else None,
        "days_of_data": len(closes),
    }

    return results

File: trading-mcp-server/src/strategies.js
// --- Indicator primitives ---

function sma(data, period) {
  if (data.length < period) return null;
  const slice = data.slice(-period);
  return slice.reduce((a, b) => a + b, 0) / period;
}

function ema(data, period) {
  if (data.length < period) return null;
  const k = 2 / (period + 1);
  let val = data.slice(0, period).reduce((a, b) => a + b, 0) / period;
  for (let i = period; i < data.length; i++) {
    val = data[i] * k + val * (1 - k);
  }
  return val;
}

function emaArray(data, period) {
  if (data.length < period) return [];
  const k = 2 / (period + 1);
  const result = [];
  let val = data.slice(0, period).reduce((a, b) => a + b, 0) / period;
  result.push(val);
  for (let i = period; i < data.length; i++) {
    val = data[i] * k + val * (1 - k);
    result.push(val);
  }
  return result;
}

function trueRange(high, low, close, i) {
  if (i === 0) return high[0] - low[0];
  return Math.max(
    high[i] - low[i],
    Math.abs(high[i] - close[i - 1]),
    Math.abs(low[i] - close[i - 1])
  );
}

function computeADX(highs, lows, closes, period = 14) {
  const len = highs.length;
  if (len < period * 2 + 1) return { adx: null, plusDI: null, minusDI: null };

  const tr = [];
  const plusDM = [];
  const minusDM = [];

  for (let i = 0; i < len; i++) {
    tr.push(trueRange(highs, lows, closes, i));
    if (i === 0) {
      plusDM.push(0);
      minusDM.push(0);
    } else {
      const upMove = highs[i] - highs[i - 1];
      const downMove = lows[i - 1] - lows[i];
      plusDM.push(upMove > downMove && upMove > 0 ? upMove : 0);
      minusDM.push(downMove > upMove && downMove > 0 ? downMove : 0);
    }
  }

  const smoothedTR = wilderSmooth(tr, period);
  const smoothedPlusDM = wilderSmooth(plusDM, period);
  const smoothedMinusDM = wilderSmooth(minusDM, period);

  const dx = [];
  let plusDI, minusDI;
  for (let i = 0; i < smoothedTR.length; i++) {
    plusDI = smoothedTR[i] > 0 ? (smoothedPlusDM[i] / smoothedTR[i]) * 100 : 0;
    minusDI = smoothedTR[i] > 0 ? (smoothedMinusDM[i] / smoothedTR[i]) * 100 : 0;
    const sum = plusDI + minusDI;
    dx.push(sum > 0 ? (Math.abs(plusDI - minusDI) / sum) * 100 : 0);
  }

  const adxArr = wilderSmooth(dx, period);
  const adx = adxArr[adxArr.length - 1];
  return { adx: r2(adx), plusDI: r2(plusDI), minusDI: r2(minusDI) };
}

function wilderSmooth(data, period) {
  if (data.length < period) return [];
  let val = data.slice(0, period).reduce((a, b) => a + b, 0);
  const result = [val];
  for (let i = period; i < data.length; i++) {
    val = val - val / period + data[i];
    result.push(val);
  }
  return result;
}

function r2(v) {
  return v == null ? null : Math.round(v * 100) / 100;
}

// --- Strategy 1: 50/200 SMA Momentum Crossover ---

function momentumSignal(closes) {
  if (closes.length < 201) return null;
  const fast = sma(closes, 50);
  const slow = sma(closes, 200);
  const prev = closes.slice(0, -1);
  const fastPrev = sma(prev, 50);
  const slowPrev = sma(prev, 200);

  let signal = "HOLD";
  if (fast > slow && fastPrev <= slowPrev) signal = "BUY";
  else if (fast < slow && fastPrev >= slowPrev) signal = "SELL";
  else if (fast > slow) signal = "BULLISH";
  else signal = "BEARISH";

  return {
    strategy: "MOMENTUM_SMA",
    signal,
    sma50: r2(fast),
    sma200: r2(slow),
    spread_pct: r2(((fast - slow) / slow) * 100),
  };
}

// --- Strategy 2: 18 EMA Breakout ---

function ema18Breakout(closes) {
  if (closes.length < 25) return null;
  const ema18 = emaArray(closes, 18);
  const offset = closes.length - ema18.length;

  const curr = closes[closes.length - 1];
  const prev = closes[closes.length - 2];
  const emaToday = ema18[ema18.length - 1];
  const emaYesterday = ema18[ema18.length - 2];

  let signal = "HOLD";
  if (prev < emaYesterday && curr > emaToday) signal = "BUY";
  else if (prev > emaYesterday && curr < emaToday) signal = "SELL";
  else if (curr > emaToday) signal = "ABOVE_EMA";
  else signal = "BELOW_EMA";

  return {
    strategy: "EMA18_BREAKOUT",
    signal,
    price: curr,
    ema18: r2(emaToday),
    distance_pct: r2(((curr - emaToday) / emaToday) * 100),
  };
}

// --- Strategy 3: Elder's Gold Impulse (EMA13 + MACD Histogram) ---

function goldImpulse(closes) {
  if (closes.length < 40) return null;
  const ema13 = emaArray(closes, 13);
  const ema12 = emaArray(closes, 12);
  const ema26 = emaArray(closes, 26);

  const macdLine = [];
  const minLen = Math.min(ema12.length, ema26.length);
  const e12 = ema12.slice(-minLen);
  const e26 = ema26.slice(-minLen);
  for (let i = 0; i < minLen; i++) {
    macdLine.push(e12[i] - e26[i]);
  }
  const signalLine = emaArray(macdLine, 9);
  const histLen = Math.min(macdLine.length, signalLine.length);
  const macdHist = [];
  for (let i = 0; i < histLen; i++) {
    macdHist.push(macdLine[macdLine.length - histLen + i] - signalLine[signalLine.length - histLen + i]);
  }

  if (macdHist.length < 2 || ema13.length < 2) return null;

  const emaRising = ema13[ema13.length - 1] > ema13[ema13.length - 2];
  const emaFalling = ema13[ema13.length - 1] < ema13[ema13.length - 2];
  const histRising = macdHist[macdHist.length - 1] > macdHist[macdHist.length - 2];
  const histFalling = macdHist[macdHist.length - 1] < macdHist[macdHist.length - 2];

  let color, signal;
  if (emaRising && histRising) {
    color = "GREEN";
    signal = "BUY";
  } else if (emaFalling && histFalling) {
    color = "RED";
    signal = "SELL";
  } else {
    color = "BLUE";
    signal = "NEUTRAL";
  }

  return {
    strategy: "GOLD_IMPULSE",
    signal,
    color,
    ema13: r2(ema13[ema13.length - 1]),
    macd_hist: r2(macdHist[macdHist.length - 1]),
  };
}

// --- Strategy 4: 80/20 Strategy (Okala) ---

function strategy8020(opens, highs, lows, closes) {
  if (closes.length < 3) return null;
  const n = closes.length;

  const bodyPrev = Math.abs(closes[n - 2] - opens[n - 2]);
  const rangePrev = highs[n - 2] - lows[n - 2];
  if (rangePrev === 0) return null;
  const bodyRatio = bodyPrev / rangePrev;

  const isBullishCandle = closes[n - 2] > opens[n - 2];
  const isBearishCandle = closes[n - 2] < opens[n - 2];
  const is8020 = bodyRatio >= 0.80;

  if (!is8020) {
    return { strategy: "STRATEGY_8020", signal: "NO_SETUP", body_ratio: r2(bodyRatio * 100) };
  }

  const todayClose = closes[n - 1];
  let signal = "SETUP_PENDING";

  if (isBullishCandle && todayClose < lows[n - 2]) {
    signal = "SELL";
  } else if (isBearishCandle && todayClose > highs[n - 2]) {
    signal = "BUY";
  } else if (isBullishCandle) {
    signal = "BEARISH_SETUP";
  } else if (isBearishCandle) {
    signal = "BULLISH_SETUP";
  }

  return {
    strategy: "STRATEGY_8020",
    signal,
    body_ratio: r2(bodyRatio * 100),
    trigger_candle: isBullishCandle ? "BULLISH" : "BEARISH",
    prev_high: highs[n - 2],
    prev_low: lows[n - 2],
  };
}

// --- Strategy 5: Volume Spike Detection ---

function volumeAnalysis(volumes, closes) {
  if (volumes.length < 21) return null;
  const avgVol = sma(volumes.slice(0, -1), 20);
  const todayVol = volumes[volumes.length - 1];
  const ratio = todayVol / avgVol;

  const priceChange = closes.length >= 2
    ? ((closes[closes.length - 1] - closes[closes.length - 2]) / closes[closes.length - 2]) * 100
    : 0;

  let signal = "NORMAL";
  if (ratio > 2.5 && priceChange > 1) signal = "BREAKOUT_VOLUME";
  else if (ratio > 2.5 && priceChange < -1) signal = "SELLOFF_VOLUME";
  else if (ratio > 1.5 && priceChange > 0.5) signal = "HIGH_BULLISH";
  else if (ratio > 1.5 && priceChange < -0.5) signal = "HIGH_BEARISH";
  else if (ratio < 0.5) signal = "DRY_UP";

  return {
    strategy: "VOLUME",
    signal,
    vol_ratio: r2(ratio),
    avg_volume: Math.round(avgVol),
    today_volume: todayVol,
    price_change_pct: r2(priceChange),
  };
}

// --- Strategy 6: ADX Regime + DI Crossover ---

function adxStrategy(highs, lows, closes) {
  const { adx, plusDI, minusDI } = computeADX(highs, lows, closes, 14);
  if (adx == null) return null;

  let regime, signal;
  if (adx > 25) {
    regime = "TRENDING";
    signal = plusDI > minusDI ? "TREND_BUY" : "TREND_SELL";
  } else if (adx < 20) {
    regime = "RANGING";
    signal = "NO_TREND";
  } else {
    regime = "WEAK_TREND";
    signal = plusDI > minusDI ? "LEAN_BULLISH" : "LEAN_BEARISH";
  }

  return {
    strategy: "ADX",
    signal,
    adx,
    plusDI,
    minusDI,
    regime,
  };
}

// --- Master Scanner ---

function scanStock(ticker, opens, highs, lows, closes, volumes) {
  const results = {
    ticker,
    last_price: closes[closes.length - 1],
    signals: {},
    actionable: [],
  };

  const adx = adxStrategy(highs, lows, closes);
  if (adx) results.signals.adx = adx;

  const mom = momentumSignal(closes);
  if (mom) results.signals.momentum = mom;

  const ema18 = ema18Breakout(closes);
  if (ema18) results.signals.ema18_breakout = ema18;

  const impulse = goldImpulse(closes);
  if (impulse) results.signals.gold_impulse = impulse;

  const eighty = strategy8020(opens, highs, lows, closes);
  if (eighty) results.signals.strategy_8020 = eighty;

  const vol = volumeAnalysis(volumes, closes);
  if (vol) results.signals.volume = vol;

  const actionableSignals = ["BUY", "SELL", "BREAKOUT_VOLUME", "SELLOFF_VOLUME"];
  for (const [key, val] of Object.entries(results.signals)) {
    if (actionableSignals.includes(val.signal)) {
      results.actionable.push({ strategy: key, signal: val.signal });
    }
  }

  // Priority score: more actionable signals + ADX confirmation = higher priority
  let score = results.actionable.length * 10;
  if (adx && adx.adx > 25 && results.actionable.length > 0) score += 5;
  if (vol && vol.vol_ratio > 2) score += 3;
  results.priority_score = score;

  return results;
}

function multiScan(stocks) {
  const results = stocks.map(({ ticker, opens, highs, lows, closes, volumes }) =>
    scanStock(ticker, opens, highs, lows, closes, volumes)
  );

  results.sort((a, b) => b.priority_score - a.priority_score);

  const actionable = results.filter((r) => r.actionable.length > 0);
  const summary = {
    total_scanned: results.length,
    actionable_count: actionable.length,
    regime_breakdown: {
      trending: results.filter((r) => r.signals.adx?.regime === "TRENDING").length,
      ranging: results.filter((r) => r.signals.adx?.regime === "RANGING").length,
      weak: results.filter((r) => r.signals.adx?.regime === "WEAK_TREND").length,
    },
  };

  return { summary, actionable, all: results };
}

export {
  multiScan,
  scanStock,
  momentumSignal,
  ema18Breakout,
  goldImpulse,
  strategy8020,
  volumeAnalysis,
  adxStrategy,
  computeADX,
  sma,
  ema,
  emaArray,
};
